rest/src/main/scala/io/udash/rest/raw/AbstractMapping.scala

```diff
@@ -0,0 +1,109 @@
+package io.udash
+package rest.raw
+
+import com.avsystem.commons._
+import io.udash.rest.raw.AbstractMapping.ConcatIterable
+
+import scala.collection.generic.CanBuildFrom
+import scala.collection.mutable
+
+sealed abstract class AbstractMapping[V] extends PartialFunction[String, V] {
+  type Self >: this.type <: AbstractMapping[V]
+
+  def entries: IIterable[(String, V)]
+  def caseSensitive: Boolean
+
+  protected def createNew(rawEntries: IIterable[(String, V)]): Self
+
+  private def normKey(key: String): String =
+    if (caseSensitive) key else key.toLowerCase
+
+  lazy val toMap: IMap[String, V] =
+    entries.iterator.map { case (name, value) => (normKey(name), value) }.toMap
+
+  def isEmpty: Boolean = entries.isEmpty
+  def nonEmpty: Boolean = entries.nonEmpty
+  def iterator: Iterator[(String, V)] = entries.iterator
+
+  def isDefinedAt(key: String): Boolean =
+    toMap.contains(normKey(key))
+  def apply(key: String): V =
+    toMap(normKey(key))
+
+  override def applyOrElse[A <: String, B >: V](key: A, default: A => B): B =
+    if (caseSensitive) toMap.applyOrElse(key, default)
```

Comment:
Wouldn't calling `normKey` every time do the same? Is it related to performance? Should JIT quite well.

Reply:
For case-insensitive, I need to rewrap the `default` into yet another lambda, which I wanted to avoid for case-sensitive.